File: src/cli/commands.ts
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { applyHandler, initHandler, revertHandler } from './handlers';

/**
 * Sets up and parses CLI commands.
 */
export function run(): void {
  yargs(hideBin(process.argv))
    .scriptName('ruler')
    .usage('$0 <command> [options]')
    .command(
      'apply',
      'Apply ruler configurations to supported AI agents',
      (y) => {
        y.option('project-root', {
          type: 'string',
          description: 'Project root directory',
          default: process.cwd(),
        });
        y.option('agents', {
          type: 'string',
          description:
            'Comma-separated list of agent identifiers: amp, copilot, claude, codex, cursor, windsurf, cline, aider, firebase, gemini-cli, junie, kilocode, opencode, crush',
        });
        y.option('config', {
          type: 'string',
          description: 'Path to TOML configuration file',
        });
        y.option('mcp', {
          type: 'boolean',
          description: 'Enable or disable applying MCP server config',
          default: true,
        });
        y.alias('mcp', 'with-mcp');
        y.option('mcp-overwrite', {
          type: 'boolean',
          description: 'Replace (not merge) the native MCP config(s)',
          default: false,
        });
        y.option('gitignore', {
          type: 'boolean',
          description:
            'Enable/disable automatic .gitignore updates (default: enabled)',
        });
        y.option('verbose', {
          type: 'boolean',
          description: 'Enable verbose logging',
          default: false,
        });
        y.alias('verbose', 'v');
        y.option('dry-run', {
          type: 'boolean',
          description: 'Preview changes without writing files',
          default: false,
        });
        y.option('local-only', {
          type: 'boolean',
          description:
            'Only search for local .ruler directories, ignore global config',
          default: false,
        });
      },
      applyHandler,
    )
    .command(
      'init',
      'Scaffold a .ruler directory with default files',
      (y) => {
        y.option('project-root', {
          type: 'string',
          description: 'Project root directory',
          default: process.cwd(),
        }).option('global', {
          type: 'boolean',
          description:
            'Initialize in global config directory (XDG_CONFIG_HOME/ruler)',
          default: false,
        });
      },
      initHandler,
    )
    .command(
      'revert',
      'Revert ruler configurations by restoring backups and removing generated files',
      (y) => {
        y.option('project-root', {
          type: 'string',
          description: 'Project root directory',
          default: process.cwd(),
        });
        y.option('agents', {
          type: 'string',
          description:
            'Comma-separated list of agent identifiers: amp, copilot, claude, codex, cursor, windsurf, cline, aider, firebase, gemini-cli, junie, kilocode, opencode, crush',
        });
        y.option('config', {
          type: 'string',
          description: 'Path to TOML configuration file',
        });
        y.option('keep-backups', {
          type: 'boolean',
          description: 'Keep backup files (.bak) after restoration',
          default: false,
        });
        y.option('verbose', {
          type: 'boolean',
          description: 'Enable verbose logging',
          default: false,
        });
        y.alias('verbose', 'v');
        y.option('dry-run', {
          type: 'boolean',
          description: 'Preview changes without actually reverting files',
          default: false,
        });
        y.option('local-only', {
          type: 'boolean',
          description:
            'Only search for local .ruler directories, ignore global config',
          default: false,
        });
      },
      revertHandler,
    )
    .demandCommand(1, 'You need to specify a command')
    .help()
    .strict()
    .parse();
}

File: rust-ruler/src/cli/commands.rs
use anyhow::Result;
use std::path::{Path, PathBuf};
use std::fs;
use directories::ProjectDirs;
use crate::types::McpStrategy;

pub fn init_command(project_root: String, global: bool) -> Result<()> {
    let ruler_dir = if global {
        // Match TypeScript: XDG_CONFIG_HOME/ruler or ~/.config/ruler
        let config_home = std::env::var("XDG_CONFIG_HOME")
            .unwrap_or_else(|_| {
                let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_string());
                format!("{}/.config", home)
            });
        PathBuf::from(config_home).join("ruler")
    } else {
        PathBuf::from(project_root).join(".ruler")
    };

    // Create the ruler directory
    fs::create_dir_all(&ruler_dir)?;

    let instructions_path = ruler_dir.join("instructions.md");
    let toml_path = ruler_dir.join("ruler.toml");
    let mcp_path = ruler_dir.join("mcp.json");

    // Check if files exist to avoid overwriting
    let file_exists = |path: &Path| path.exists();

    let default_instructions = r#"# Ruler Instructions

These are your centralised AI agent instructions.
Add your coding guidelines, style guides, and other project-specific context here.

Ruler will concatenate all .md files in this directory (and its subdirectories)
and apply them to your configured AI coding agents.
"#;

    let default_toml = r#"# Ruler Configuration File
# See https://ai.intellectronica.net/ruler for documentation.

# To specify which agents are active by default when --agents is not used,
# uncomment and populate the following line. If omitted, all agents are active.
# default_agents = ["copilot", "claude"]

# --- Agent Specific Configurations ---
# You can enable/disable agents and override their default output paths here.
# Use lowercase agent identifiers: amp, copilot, claude, codex, cursor, windsurf, cline, aider, kilocode

# [agents.copilot]
# enabled = true
# output_path = ".github/copilot-instructions.md"

# [agents.claude]
# enabled = true
# output_path = "CLAUDE.md"

# [agents.codex]
# enabled = true
# output_path = "AGENTS.md"

# [agents.cursor]
# enabled = true
# output_path = ".cursor/rules/ruler_cursor_instructions.mdc"

# [agents.windsurf]
# enabled = true
# output_path = ".windsurf/rules/ruler_windsurf_instructions.md"

# [agents.cline]
# enabled = true
# output_path = ".clinerules"

# [agents.aider]
# enabled = true
# output_path_instructions = "ruler_aider_instructions.md"
# output_path_config = ".aider.conf.yml"

# [agents.firebase]
# enabled = true
# output_path = ".idx/airules.md"

# [agents.gemini-cli]
# enabled = true

# [agents.kilocode]
# enabled = true
# output_path = ".kilocode/rules/ruler_kilocode_instructions.md"
"#;

    let default_mcp_json = r#"{
  "mcpServers": {
    "example": {
      "type": "stdio",
      "command": "node",
      "args": ["/path/to/mcp-server.js"],
      "env": {
        "NODE_ENV": "production"
      }
    }
  }
}
"#;

    // Create instructions.md if it doesn't exist
    if !file_exists(&instructions_path) {
        fs::write(&instructions_path, default_instructions)?;
        println!("[ruler] Created {}", instructions_path.display());
    } else {
        println!("[ruler] instructions.md already exists, skipping");
    }

    // Create ruler.toml if it doesn't exist
    if !file_exists(&toml_path) {
        fs::write(&toml_path, default_toml)?;
        println!("[ruler] Created {}", toml_path.display());
    } else {
        println!("[ruler] ruler.toml already exists, skipping");
    }

    // Create mcp.json if it doesn't exist
    if !file_exists(&mcp_path) {
        fs::write(&mcp_path, default_mcp_json)?;
        println!("[ruler] Created {}", mcp_path.display());
    } else {
        println!("[ruler] mcp.json already exists, skipping");
    }

    Ok(())
}

pub fn apply_command(
    project_root: String,
    agents: Option<Vec<String>>,
    config: Option<String>,
    mcp_enabled: bool,
    mcp_strategy: Option<McpStrategy>,
    gitignore_enabled: Option<bool>,
    verbose: bool,
    dry_run: bool,
    local_only: bool,
) -> Result<()> {
    // Simple implementation - just print what would be done for now
    let action_prefix = if dry_run { "[ruler:dry-run]" } else { "[ruler]" };
    
    if verbose {
        println!("Loading configuration from project root: {}", project_root);
    }
    
    println!("{} Apply command called with project_root: {}", action_prefix, project_root);
    
    if let Some(ref agents_list) = agents {
        if verbose {
            println!("Agents specified: {:?}", agents_list);
        }
        for agent in agents_list {
            println!("{} Would apply rules to agent: {}", action_prefix, agent);
        }
    } else {
        println!("{} Would apply rules to all enabled agents", action_prefix);
    }
    
    if !dry_run {
        println!("✓ Configuration applied successfully (placeholder)");
    } else {
        println!("✓ Dry run completed - no files were modified");
    }
    
    Ok(())
}pub fn revert_command(
    project_root: String,
    agents: Option<Vec<String>>,
    config: Option<String>,
    keep_backups: bool,
    verbose: bool,
    dry_run: bool,
    local_only: bool,
) -> Result<()> {
    // TODO: Implement revert command  
    println!("Revert command not yet implemented");
    Ok(())
}
